File: backend/src/main/java/kr/co/programmers/collabond/api/recruit/application/RecruitPostService.java
package kr.co.programmers.collabond.api.recruit.application;

import kr.co.programmers.collabond.api.image.domain.Image;
import kr.co.programmers.collabond.api.profile.application.ProfileService;
import kr.co.programmers.collabond.api.profile.domain.Profile;
import kr.co.programmers.collabond.api.recruit.domain.RecruitPost;
import kr.co.programmers.collabond.api.recruit.domain.RecruitPostStatus;
import kr.co.programmers.collabond.api.recruit.domain.dto.RecruitPostRequestDto;
import kr.co.programmers.collabond.api.recruit.domain.dto.RecruitPostResponseDto;
import kr.co.programmers.collabond.api.recruit.infrastructure.RecruitPostRepository;
import kr.co.programmers.collabond.api.recruit.interfaces.RecruitPostMapper;
import kr.co.programmers.collabond.api.user.application.UserService;
import kr.co.programmers.collabond.api.user.domain.User;
import kr.co.programmers.collabond.core.auth.oauth2.OAuth2UserInfo;
import kr.co.programmers.collabond.shared.exception.ErrorCode;
import kr.co.programmers.collabond.shared.exception.ForbiddenException;
import kr.co.programmers.collabond.shared.exception.InvalidException;
import kr.co.programmers.collabond.shared.exception.NotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class RecruitPostService {

    private final ProfileService profileService;
    private final UserService userService;

    private final RecruitPostRepository recruitPostRepository;

    // 모집글 작성
    @Transactional
    public RecruitPostResponseDto createRecruitPost(RecruitPostRequestDto request,
                                                    OAuth2UserInfo userInfo) {

        Profile profile = profileService.findByProfileId(request.getProfileId());

        User loginUser = userService.findByProviderId(userInfo.getUsername());

        if (!profile.getUser().getId().equals(loginUser.getId())) {
            throw new ForbiddenException();
        }

        RecruitPost post = RecruitPostMapper.toEntity(request, profile);
        String imgPath = getProfileImgName(post);

        return RecruitPostMapper.toResponseDto(recruitPostRepository.save(post), imgPath);
    }

    // 모집글 수정
    @Transactional
    public RecruitPostResponseDto updateRecruitPost(Long recruitmentId,
                                                    RecruitPostRequestDto request,
                                                    OAuth2UserInfo userInfo) {

        RecruitPost post = recruitPostRepository.findById(recruitmentId)
                .orElseThrow(() -> new NotFoundException(ErrorCode.RECRUIT_NOT_FOUND));

        // 소프트 삭제된 게시글은 수정할 수 없습니다.
        if (post.getDeletedAt() != null) {
            throw new InvalidException(ErrorCode.REMOVED_RECRUIT_POST);
        }

        User loginUser = userService.findByProviderId(userInfo.getUsername());

        if (!post.getProfile().getUser().getId().equals(loginUser.getId())) {
            throw new ForbiddenException();
        }

        post.update(request.getTitle(),
                request.getDescription(),
                RecruitPostStatus.valueOf(request.getStatus()),
                request.getDeadline());

        String imgPath = getProfileImgName(post);

        return RecruitPostMapper.toResponseDto(post, imgPath);
    }

    // 모집글 삭제 (소프트 삭제)
    @Transactional
    public void deleteRecruitPost(Long recruitmentId, OAuth2UserInfo userInfo) {

        RecruitPost post = recruitPostRepository.findById(recruitmentId)
                .orElseThrow(() -> new NotFoundException(ErrorCode.RECRUIT_NOT_FOUND));

        User loginUser = userService.findByProviderId(userInfo.getUsername());

        if (!post.getProfile().getUser().getId().equals(loginUser.getId())) {
            throw new ForbiddenException();
        }

        recruitPostRepository.delete(post);
    }

    // 모집글 목록 조회
    @Transactional(readOnly = true)
    public Page<RecruitPostResponseDto> getAllRecruitPosts(RecruitPostStatus status,
                                                           String sort,
                                                           Pageable pageable) {

        Page<RecruitPost> posts = status != null
                ? recruitPostRepository.findByStatus(status, pageable)
                : recruitPostRepository.findAll(pageable);

        return posts.map(recruitPost -> {
            String imgPath = getProfileImgName(recruitPost);
            return RecruitPostMapper.toResponseDto(recruitPost, imgPath);
        });
    }

    // 회원이 작성한 모집글 조회
    @Transactional(readOnly = true)
    public Page<RecruitPostResponseDto> getRecruitPostsByUser(Long userId, Pageable pageable) {
        return recruitPostRepository.findByUserId(userId, pageable)
                .map(recruitPost -> {
                    String imgPath = getProfileImgName(recruitPost);
                    return RecruitPostMapper.toResponseDto(recruitPost, imgPath);
                });
    }

    // 프로필이 작성한 모집글 조회
    @Transactional(readOnly = true)
    public Page<RecruitPostResponseDto> getRecruitPostByProfile(Long profileId, Pageable pageable) {
        return recruitPostRepository.findByProfileId(profileId, pageable)
                .map(recruitPost -> {
                    String imgPath = getProfileImgName(recruitPost);
                    return RecruitPostMapper.toResponseDto(recruitPost, imgPath);
                });
    }

    @Transactional
    public RecruitPost findByRecruitmentId(Long recruitmentId) {
        return recruitPostRepository.findById(recruitmentId)
                .orElseThrow(() -> new NotFoundException(ErrorCode.RECRUIT_NOT_FOUND));
    }

    private String getProfileImgName(RecruitPost recruitPost) {
        Image profileImage = recruitPost.getProfile().getImages().stream()
                .filter(i -> i.getType().equals("PROFILE"))
                .findAny()
                .orElse(null);
        return profileImage.getFile().getSavedName();
    }

    // 단건 조회 기능 - 5/23 수정
    public RecruitPostResponseDto getRecruitPostById(Long recruitmentId) {
        RecruitPost recruitPost = recruitPostRepository.findById(recruitmentId)
                .orElseThrow(() -> new NotFoundException(ErrorCode.RECRUIT_NOT_FOUND));

        String imgPath = getProfileImgName(recruitPost);
        return RecruitPostMapper.toResponseDto(recruitPost, imgPath);
    }
}

File: backend/src/main/java/kr/co/programmers/collabond/api/recruit/application/RecruitPostService.kt
package kr.co.programmers.collabond.api.recruit.application

import kr.co.programmers.collabond.api.profile.application.ProfileService
import kr.co.programmers.collabond.api.recruit.domain.RecruitPost
import kr.co.programmers.collabond.api.recruit.domain.RecruitPostStatus
import kr.co.programmers.collabond.api.recruit.domain.dto.Requests
import kr.co.programmers.collabond.api.recruit.domain.dto.Responses
import kr.co.programmers.collabond.api.recruit.infrastructure.RecruitPostRepository
import kr.co.programmers.collabond.api.recruit.interfaces.RecruitPostMapper
import kr.co.programmers.collabond.api.user.application.UserService
import kr.co.programmers.collabond.core.auth.oauth2.OAuth2UserInfo
import kr.co.programmers.collabond.shared.exception.ErrorCode
import kr.co.programmers.collabond.shared.exception.ForbiddenException
import kr.co.programmers.collabond.shared.exception.InvalidException
import kr.co.programmers.collabond.shared.exception.NotFoundException
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class RecruitPostService(
    private val profileService: ProfileService,
    private val userService: UserService,
    private val recruitPostRepository: RecruitPostRepository
) {

    @Transactional
    fun createRecruitPost(
        request: Requests,
        userInfo: OAuth2UserInfo
    ): Responses {
        val profile = profileService.findByProfileId(request.profileId)
        val loginUser = userService.findByProviderId(userInfo.username)

        if (profile.user.id != loginUser.id) throw ForbiddenException()

        val post = RecruitPostMapper.toEntity(request, profile)
        val savedPost = recruitPostRepository.save(post)

        return RecruitPostMapper.toResponseDto(savedPost);
    }

    @Transactional
    fun updateRecruitPost(
        recruitmentId: Long,
        request: Requests,
        userInfo: OAuth2UserInfo
    ): Responses {
        val post = findRecruitPostById(recruitmentId)

        // 소프트 삭제된 게시글은 수정할 수 없습니다.
        if (post.deletedAt != null) throw InvalidException(ErrorCode.REMOVED_RECRUIT_POST)

        val loginUser = userService.findByProviderId(userInfo.username)
        if (post.profile?.user?.id != loginUser.id) throw ForbiddenException()

        post.update(
            title = request.title,
            description = request.description,
            status = request.status?.let(RecruitPostStatus::valueOf),
            deadline = request.deadline
        )

        return RecruitPostMapper.toResponseDto(post)
    }

    @Transactional
    fun deleteRecruitPost(recruitmentId: Long, userInfo: OAuth2UserInfo) {
        val post = findRecruitPostById(recruitmentId)
        val loginUser = userService.findByProviderId(userInfo.username)

        if (post.profile?.user?.id != loginUser.id) throw ForbiddenException()

        recruitPostRepository.delete(post)
    }

    @Transactional(readOnly = true)
    fun getAllRecruitPosts(
        status: RecruitPostStatus? = null,
        sort: String? = null,
        pageable: Pageable
    ): Page<Responses> {
        val posts = status?.let { recruitPostRepository.findByStatus(it, pageable) }
            ?: recruitPostRepository.findAll(pageable)

        return posts.map { recruitPost ->
            RecruitPostMapper.toResponseDto(recruitPost)
        }
    }

    @Transactional(readOnly = true)
    fun getRecruitPostsByUser(userId: Long, pageable: Pageable): Page<Responses> =
        recruitPostRepository.findByUserId(userId, pageable)
            .map { recruitPost ->
                RecruitPostMapper.toResponseDto(recruitPost)
            }

    @Transactional(readOnly = true)
    fun getRecruitPostByProfile(profileId: Long, pageable: Pageable): Page<Responses> =
        recruitPostRepository.findByProfileId(profileId, pageable)
            .map { recruitPost ->
                RecruitPostMapper.toResponseDto(recruitPost)
            }

    @Transactional(readOnly = true)
    fun findByRecruitmentId(recruitmentId: Long): RecruitPost = findRecruitPostById(recruitmentId)

    @Transactional(readOnly = true)
    fun getRecruitPostById(recruitmentId: Long): Responses {
        val recruitPost = findRecruitPostById(recruitmentId)
        return RecruitPostMapper.toResponseDto(recruitPost)
    }

    private fun findRecruitPostById(recruitmentId: Long): RecruitPost =
        recruitPostRepository.findByIdOrNull(recruitmentId)
            ?: throw NotFoundException(ErrorCode.RECRUIT_NOT_FOUND)

    private fun RecruitPost.getProfileImgName(): String =
        profile?.images?.firstOrNull { it.type == "PROFILE" }?.file?.savedName.orEmpty()
}